### ISSUE TITLE
Retry policy when a worker crashes: a hook missing?

### ISSUE THREAD

**Issue by mitar** [ray-project/ray#2635]
So currently if a worker crashes ray task running on that worker is automatically retried. Some things about that:
* I have not seen any way to control how many times this is retried? If my Python code calls into some C code which segfaults, I do not want it to be retrying all the time.
* How do I know as a caller that this happens? I have not seen anything about this available in `ray.get(...)`. I just see messages printed to my stdout as a caller, but not really something I can detect programmatically?
* This assumes that functions are pure or at least idempotent. But what if they are not? Then calling a function again might break things and get things into an undefined state (or into using an undefined state left over from the previous run).

So, maybe what is missing is a hook which is called once a worker dies. Which by default marks the task for retrying, but it can also do something else. Like cleanup broken outside state used by the task, or decide not to run and propagate an error to the caller. I would not mind even that `ray.get(...)` simply aborts with an exception saying "worker terminated", do whatever you want. But ideally, I would prefer to have a hook I can register to handle even task I am not `ray.get`-ting.

**Comment by stephanie-wang** [comment]
Hi @mitar, in response for your first two bullet points, one thing we're considering for this is propagating the error as an exception that would be thrown by `ray.get(object_id)`, and then providing a Ray method (e.g., `ray.retry(object_id)`) to retry the task that produced the error. Would something like that work for you? I think providing the user with a hook is a good idea but might be problematic since it locks the user into using callbacks. But please let us know if you have any suggestions or thoughts on what the retry API should look like.

**Comment by mitar** [comment]
Yes, `ray.get` and `ray.retry` would be satisfactory. I can always make a wrapper around which then does what I would do in a callback in a general way.

**Comment by stephanie-wang** [comment]
Okay, great! It may be several weeks before something like this would work end-to-end, but it is definitely on our list. The first step is to propagate the backend error to Python. The second step is to provide an experimental `ray.retry` method.

**Comment by mitar** [comment]
Why? Has this been implemented?

**Comment by mitar** [comment]
Unstale.

**Comment by mitar** [comment]
Unstale.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
